// react-styleguidist/lib/rsg-components/Editor/Editor.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';
import debounce from 'lodash/debounce';
import { UnControlled as CodeMirror } from 'react-codemirror2';
import 'codemirror/mode/jsx/jsx';

// We’re explicitly specifying Webpack loaders here so we could skip specifying them in Webpack configuration.
// That way we could avoid clashes between our loaders and user loaders.
// eslint-disable-next-line import/no-unresolved
require('!!../../../loaders/style-loader!../../../loaders/css-loader!codemirror/lib/codemirror.css');
// eslint-disable-next-line import/no-unresolved
require('!!../../../loaders/style-loader!../../../loaders/css-loader!rsg-codemirror-theme.css');

var UPDATE_DELAY = 10;

var styles = function styles(_ref) {
	var fontFamily = _ref.fontFamily,
	    space = _ref.space,
	    fontSize = _ref.fontSize;
	return {
		root: {
			'& .CodeMirror': {
				isolate: false,
				fontFamily: fontFamily.monospace,
				height: 'auto',
				padding: [[space[1], space[2]]],
				fontSize: fontSize.small
			},
			'& .CodeMirror pre': {
				isolate: false,
				padding: 0
			},
			'& .CodeMirror-scroll': {
				isolate: false,
				height: 'auto',
				overflowY: 'hidden',
				overflowX: 'auto'
			},
			'& .cm-error': {
				isolate: false,
				background: 'none'
			}
		}
	};
};

export var Editor = function (_Component) {
	_inherits(Editor, _Component);

	function Editor() {
		_classCallCheck(this, Editor);

		var _this = _possibleConstructorReturn(this, (Editor.__proto__ || Object.getPrototypeOf(Editor)).call(this));

		_this.handleChange = debounce(_this.handleChange.bind(_this), UPDATE_DELAY);
		return _this;
	}

	_createClass(Editor, [{
		key: 'shouldComponentUpdate',
		value: function shouldComponentUpdate(nextProps) {
			return !!(this.getEditorConfig(nextProps).readOnly && nextProps.code !== this.props.code);
		}
	}, {
		key: 'getEditorConfig',
		value: function getEditorConfig(props) {
			return _extends({}, this.context.config.editorConfig, props.editorConfig);
		}
	}, {
		key: 'handleChange',
		value: function handleChange(editor, metadata, newCode) {
			var onChange = this.props.onChange;

			if (onChange) {
				onChange(newCode);
			}
		}
	}, {
		key: 'render',
		value: function render() {
			var _props = this.props,
			    code = _props.code,
			    classes = _props.classes;

			return React.createElement(CodeMirror, {
				className: classes.root,
				value: code,
				onChange: this.handleChange,
				options: this.getEditorConfig(this.props)
			});
		}
	}]);

	return Editor;
}(Component);

Editor.propTypes = {
	code: PropTypes.string.isRequired,
	onChange: PropTypes.func,
	editorConfig: PropTypes.object,
	classes: PropTypes.object.isRequired
};
Editor.contextTypes = {
	config: PropTypes.object.isRequired
};
export default Styled(styles)(Editor);
