// @sanity/structure/lib/Editor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.editorWithInitialValueTemplate = editorWithInitialValueTemplate;
exports.EditorBuilder = void 0;

var _camelCase2 = _interopRequireDefault(require("lodash/camelCase"));

var _initialValueTemplates = require("@sanity/initial-value-templates");

var _SerializeError = require("./SerializeError");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class EditorBuilder {
  constructor(spec) {
    _defineProperty(this, "spec", void 0);

    this.spec = spec ? spec : {};
  }

  id(id) {
    return this.clone({
      id
    });
  }

  getId() {
    return this.spec.id;
  }

  title(title) {
    return this.clone({
      title,
      id: this.spec.id || (0, _camelCase2.default)(title)
    });
  }

  getTitle() {
    return this.spec.title;
  }

  documentId(documentId) {
    // Let's try to be a bit helpful and assign an ID from document ID if none is specified
    var paneId = this.spec.id || documentId;
    return this.clone({
      id: paneId,
      options: _objectSpread({}, this.spec.options || {}, {
        id: documentId
      })
    });
  }

  getDocumentId() {
    return this.spec.options && this.spec.options.id;
  }

  schemaType(documentType) {
    return this.clone({
      options: _objectSpread({}, this.spec.options || {}, {
        type: typeof documentType === 'string' ? documentType : documentType.name
      })
    });
  }

  getSchemaType() {
    return this.spec.options && this.spec.options.type;
  }

  initialValueTemplate(templateId, parameters) {
    return this.clone({
      options: _objectSpread({}, this.spec.options || {}, {
        template: templateId,
        templateParameters: parameters
      })
    });
  }

  getInitalValueTemplate() {
    return this.spec.options && this.spec.options.template;
  }

  getInitialValueTemplateParameters() {
    return this.spec.options && this.spec.options.templateParameters;
  }

  serialize() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    },
        _ref$path = _ref.path,
        path = _ref$path === void 0 ? [] : _ref$path,
        index = _ref.index,
        hint = _ref.hint;

    var urlId = path[index || path.length - 1]; // Try to grab document ID / editor ID from URL if not defined

    var id = this.spec.id || urlId && "".concat(urlId) || '';

    var options = _objectSpread({
      id,
      type: undefined,
      template: undefined,
      templateParameters: undefined
    }, this.spec.options);

    if (typeof id !== 'string' || !id) {
      throw new _SerializeError.SerializeError('`id` is required for editor nodes', path, index, hint).withHelpUrl(_SerializeError.HELP_URL.ID_REQUIRED);
    }

    if (!options || !options.id) {
      throw new _SerializeError.SerializeError('document id (`id`) is required for editor nodes', path, id, hint).withHelpUrl(_SerializeError.HELP_URL.DOCUMENT_ID_REQUIRED);
    }

    return _objectSpread({}, this.spec, {
      id,
      type: 'document',
      options: getEditorOptions(options)
    });
  }

  clone() {
    var withSpec = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var builder = new EditorBuilder();

    var options = _objectSpread({}, this.spec.options || {}, {}, withSpec.options || {});

    builder.spec = _objectSpread({}, this.spec, {}, withSpec, {
      options
    });
    return builder;
  }

}

exports.EditorBuilder = EditorBuilder;

function getEditorOptions(spec) {
  var opts = {
    id: spec.id || '',
    type: spec.type || '*'
  };

  if (spec.template) {
    opts.template = spec.template;
  }

  if (spec.templateParameters) {
    opts.templateParameters = spec.templateParameters;
  }

  return opts;
}

function editorWithInitialValueTemplate(templateId, parameters) {
  var template = (0, _initialValueTemplates.getTemplateById)(templateId);

  if (!template) {
    throw new Error("Template with ID \"".concat(templateId, "\" not defined"));
  }

  return new EditorBuilder().schemaType(template.schemaType).initialValueTemplate(templateId, parameters);
}
