// @sanity/form-builder/lib/inputs/ImageToolInput/ImageToolInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

var _default = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _imagetool = _interopRequireDefault(require("@sanity/imagetool"));

var _HotspotImage = _interopRequireDefault(require("@sanity/imagetool/HotspotImage"));

var _imageLoader = _interopRequireDefault(require("part:@sanity/components/utilities/image-loader"));

var _constants = require("@sanity/imagetool/constants");

var _ImageToolInput = _interopRequireDefault(require("./styles/ImageToolInput.css"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PREVIEW_ASPECT_RATIOS = [['Portrait', 9 / 16], ['Square', 1], ['Landscape', 16 / 9], ['Panorama', 4]];

class ImageToolInput extends _react.default.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "handleChangeEnd", () => {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          readOnly = _this$props.readOnly,
          type = _this$props.type;
      var value = this.state.value; // For backwards compatibility, where hotspot/crop might not have a named type yet

      var cropField = type.fields.find(field => field.name === 'crop' && field.type.name !== 'object');
      var hotspotField = type.fields.find(field => field.name === 'hotspot' && field.type.name !== 'object');

      if (!readOnly) {
        var crop = cropField ? _objectSpread({
          _type: cropField.type.name
        }, value.crop) : value.crop;
        var hotspot = hotspotField ? _objectSpread({
          _type: hotspotField.type.name
        }, value.hotspot) : value.hotspot;
        onChange(_PatchEvent.default.from([(0, _PatchEvent.set)(crop, ['crop']), (0, _PatchEvent.set)(hotspot, ['hotspot'])]));
      }

      this.setState({
        value: this.props.value
      });
    });

    _defineProperty(this, "handleChange", nextValue => {
      this.setState({
        value: nextValue
      });
    });

    this.state = {
      value: props.value
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    this.setState({
      value: nextProps.value
    });
  }

  render() {
    var _this$props2 = this.props,
        imageUrl = _this$props2.imageUrl,
        level = _this$props2.level,
        readOnly = _this$props2.readOnly;
    var value = this.state.value;
    return _react.default.createElement("div", {
      className: _ImageToolInput.default.root
    }, _react.default.createElement(_default.default, {
      legend: "Hotspot and crop",
      level: level
    }, _react.default.createElement("div", {
      className: _ImageToolInput.default.wrapper
    }, _react.default.createElement("div", {
      className: _ImageToolInput.default.imageToolContainer
    }, _react.default.createElement(_imagetool.default, {
      value: value,
      src: imageUrl,
      readOnly: readOnly,
      onChangeEnd: this.handleChangeEnd,
      onChange: this.handleChange
    })), _react.default.createElement("div", {
      className: _ImageToolInput.default.previewsContainer
    }, _react.default.createElement("h2", null, "Preview"), _react.default.createElement("div", {
      className: _ImageToolInput.default.previews
    }, PREVIEW_ASPECT_RATIOS.map((_ref) => {
      var _ref2 = _slicedToArray(_ref, 2),
          title = _ref2[0],
          ratio = _ref2[1];

      return _react.default.createElement("div", {
        key: ratio,
        className: _ImageToolInput.default.preview
      }, _react.default.createElement("h4", null, title), _react.default.createElement("div", {
        className: _ImageToolInput.default.previewImage
      }, _react.default.createElement(_imageLoader.default, {
        src: imageUrl
      }, (_ref3) => {
        var image = _ref3.image,
            error = _ref3.error;
        return error ? _react.default.createElement("span", null, "Unable to load image: ", error.message) : _react.default.createElement(_HotspotImage.default, {
          aspectRatio: ratio,
          src: image.src,
          srcAspectRatio: image.width / image.height,
          hotspot: value.hotspot || _constants.DEFAULT_HOTSPOT,
          crop: value.crop || _constants.DEFAULT_CROP
        });
      })));
    }))))));
  }

}

exports.default = ImageToolInput;
