// @sanity/components/lib/badges/DefaultBadge.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _DefaultBadgeModules = _interopRequireDefault(require("./styles/DefaultBadge.modules.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DefaultBadge extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        color = _this$props.color,
        inverted = _this$props.inverted,
        children = _this$props.children,
        title = _this$props.title,
        faded = _this$props.faded;
    return _react.default.createElement("span", {
      className: inverted ? _DefaultBadgeModules.default.inverted : _DefaultBadgeModules.default.default,
      "data-color": color,
      "data-faded": faded,
      title: title
    }, children);
  }

}

exports.default = DefaultBadge;

_defineProperty(DefaultBadge, "propTypes", {
  color: _propTypes.default.oneOf([undefined, 'success', 'warning', 'danger', 'info', 'neutral']),
  inverted: _propTypes.default.bool,
  faded: _propTypes.default.bool,
  children: _propTypes.default.node,
  title: _propTypes.default.string
});

_defineProperty(DefaultBadge, "defaultProps", {
  color: undefined,
  inverted: false
});
