// @sanity/base/lib/components/MissingProjectConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MissingProjectConfig;

var _react = _interopRequireDefault(require("react"));

var _configSanity = _interopRequireDefault(require("config:sanity"));

var _content = _interopRequireDefault(require("part:@sanity/components/dialogs/content?"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen?"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function MissingProjectConfig() {
  var root = _configSanity.default.root,
      project = _configSanity.default.project,
      plugins = _configSanity.default.plugins;

  var _ref = _configSanity.default.api || {},
      dataset = _ref.dataset,
      projectId = _ref.projectId;

  var api = {
    projectId: projectId || 'myProjectId',
    dataset: dataset || 'myDatasetName'
  };
  var desiredConfig = {
    root,
    project,
    api,
    plugins
  };
  var missing = [!projectId && '"projectId"', !dataset && '"dataset"'].filter(Boolean);
  return _react.default.createElement(_fullscreen.default, {
    color: "default",
    title: "Project details missing",
    isOpen: true,
    centered: true
  }, _react.default.createElement(_content.default, {
    size: "medium",
    padding: "none"
  }, _react.default.createElement("p", null, "The ", _react.default.createElement("code", null, "sanity.json"), " file in your studio folder seems to be missing the", ' ', missing.join(' and '), " configuration ", missing.length > 1 ? 'options ' : 'option ', "under the ", _react.default.createElement("code", null, "api"), " key."), _react.default.createElement("p", null, "A valid ", _react.default.createElement("code", null, "sanity.json"), " file looks something like the following:"), _react.default.createElement("pre", null, _react.default.createElement("code", null, highlightMissing(JSON.stringify(desiredConfig, null, 2))))));
} // eslint-disable-next-line react/no-multi-comp


function highlightMissing(jsonConfig) {
  var parts = jsonConfig.split(/("dataset": "myDatasetName"|"projectId": "myProjectId")/).reduce((els, part, i) => {
    if (i % 2 === 0) {
      return els.concat(part);
    }

    return els.concat(_react.default.createElement("strong", {
      key: part
    }, part));
  }, []);
  return _react.default.createElement(_react.default.Fragment, null, parts);
}
