// @sanity/default-layout/lib/components/InitialValueTemplateError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = InitialValueTemplateError;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _content = _interopRequireDefault(require("part:@sanity/components/dialogs/content"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function InitialValueTemplateError(_ref) {
  var errors = _ref.errors;
  return _react.default.createElement(_fullscreen.default, {
    color: "danger",
    title: "Initial value template error",
    isOpen: true,
    centered: true
  }, _react.default.createElement(_content.default, {
    size: "medium",
    padding: "none"
  }, _react.default.createElement("p", null, "Failed to load initial value templates:"), errors.map((error, i) => _react.default.createElement("p", {
    key: error.message
  }, _react.default.createElement("code", null, error.message)))));
}

InitialValueTemplateError.propTypes = {
  errors: _propTypes.default.arrayOf(_propTypes.default.shape({
    message: _propTypes.default.string.isRequired
  })).isRequired
};
