// @sanity/desk-tool/lib/pane/EditorStatusBadge.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default = _interopRequireDefault(require("part:@sanity/components/badges/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class EditorStatusBadge extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        isDraft = _this$props.isDraft,
        isPublished = _this$props.isPublished,
        title = _this$props.title,
        liveEdit = _this$props.liveEdit;
    return _react.default.createElement(_react.default.Fragment, null, liveEdit ? _react.default.createElement(_default.default, {
      color: "danger"
    }, "Live") : _react.default.createElement(_react.default.Fragment, null, !isDraft && !isPublished && _react.default.createElement(_default.default, {
      inverted: true,
      faded: true
    }, "Draft"), isPublished && _react.default.createElement(_default.default, {
      color: "success",
      title: title
    }, "Published"), isDraft && _react.default.createElement(_default.default, {
      inverted: true,
      color: "neutral"
    }, "Draft")));
  }

}

exports.default = EditorStatusBadge;

_defineProperty(EditorStatusBadge, "propTypes", {
  isDraft: _propTypes.default.bool,
  isPublished: _propTypes.default.bool,
  title: _propTypes.default.string,
  liveEdit: _propTypes.default.bool
});
