// @sanity/desk-tool/lib/pane/History.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _scroll = _interopRequireDefault(require("scroll"));

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/snackbar/default"));

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _HistoryItem = _interopRequireDefault(require("./HistoryItem"));

var _History = _interopRequireDefault(require("./styles/History.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var maybe = (val, fn) => val && fn(val);

class History extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      headerShadowOpacity: 0
    });

    _defineProperty(this, "_listElement", _react.default.createRef());

    _defineProperty(this, "handleListScroll", event => {
      var scrollTop = event.target.scrollTop;
      this.setState({
        headerShadowOpacity: Math.min(scrollTop, 50) / 100
      });
    });

    _defineProperty(this, "handleNewCurrentEvent", () => {
      if (this._listElement && this._listElement.current) {
        _scroll.default.top(this._listElement.current, 0);
      }
    });
  }

  componentDidMount() {
    if (this._listElement && this._listElement.current) {
      this._listElement.current.addEventListener('scroll', this.handleListScroll, {
        passive: true
      });
    }
  }

  componentWillUnmount() {
    if (this._listElement && this._listElement.current) {
      this._listElement.current.removeEventListener('scroll', this.handleListScroll);
    }
  }

  componentDidUpdate(prevProps) {
    var _this$props = this.props,
        selectedRev = _this$props.selectedRev,
        events = _this$props.events,
        historyValue = _this$props.historyValue;

    if (prevProps.historyValue && !historyValue && events[0].rev === selectedRev) {
      this.handleNewCurrentEvent();
    }
  }

  render() {
    var _this$props2 = this.props,
        onClose = _this$props2.onClose,
        events = _this$props2.events,
        onItemSelect = _this$props2.onItemSelect,
        selectedEvent = _this$props2.selectedEvent,
        isLoading = _this$props2.isLoading,
        error = _this$props2.error;
    var headerShadowOpacity = this.state.headerShadowOpacity;
    return _react.default.createElement("div", {
      className: _History.default.root
    }, _react.default.createElement("div", {
      className: _History.default.header,
      style: {
        boxShadow: "0 0px 2px rgba(0, 0, 0, ".concat(headerShadowOpacity, ")")
      }
    }, "History", _react.default.createElement(_default.default, {
      onClick: onClose,
      title: "Close",
      icon: _closeIcon.default,
      bleed: true,
      kind: "simple",
      className: _History.default.closeButton
    })), isLoading && _react.default.createElement(_spinner.default, {
      center: true,
      message: "Loading history"
    }), error && _react.default.createElement("p", null, "Could not load history"), _react.default.createElement("div", {
      className: _History.default.list,
      ref: this._listElement
    }, !(isLoading || error) && events.map((event, i) => _react.default.createElement(_HistoryItem.default, _extends({}, event, {
      key: event.rev,
      onClick: () => onItemSelect(events[i]),
      isSelected: event === selectedEvent,
      isCurrentVersion: i === 0,
      onSelectPrev: () => maybe(events[i - 1], onItemSelect),
      onSelectNext: () => maybe(events[i + 1], onItemSelect)
    })))), error && _react.default.createElement(_default2.default, {
      kind: "error",
      isPersisted: true,
      title: error
    }));
  }

}

exports.default = History;

_defineProperty(History, "propTypes", {
  events: _propTypes.default.arrayOf(_propTypes.default.object),
  onClose: _propTypes.default.func,
  documentId: _propTypes.default.string,
  onItemSelect: _propTypes.default.func,
  isLoading: _propTypes.default.bool,
  lastEdited: _propTypes.default.object,
  errorMessage: _propTypes.default.string,
  selectedEvent: _propTypes.default.object,
  selectedRev: _propTypes.default.string,
  historyValue: _propTypes.default.object,
  error: _propTypes.default.object
});

_defineProperty(History, "defaultProps", {
  isLoading: true
});
