// @sanity/structure/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SerializeError", {
  enumerable: true,
  get: function get() {
    return _SerializeError.SerializeError;
  }
});
exports.StructureBuilder = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _List = require("./List");

var _documentTypeListItems = require("./documentTypeListItems");

var _MenuItem = require("./MenuItem");

var _ListItem = require("./ListItem");

var _MenuItemGroup = require("./MenuItemGroup");

var _DocumentList = require("./DocumentList");

var _Editor = require("./Editor");

var _SerializeError = require("./SerializeError");

var _Component = require("./Component");

var _DocumentListItem = require("./DocumentListItem");

var _InitialValueTemplateItem = require("./InitialValueTemplateItem");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StructureBuilder = {
  defaults: getDefaultStructure,
  documentTypeList: _documentTypeListItems.getDocumentTypeList,
  documentTypeListItem: _documentTypeListItems.getDocumentTypeListItem,
  documentTypeListItems: _documentTypeListItems.getDocumentTypeListItems,
  list: spec => new _List.ListBuilder(spec),
  listItem: spec => new _ListItem.ListItemBuilder(spec),
  menuItem: spec => new _MenuItem.MenuItemBuilder(spec),
  menuItemGroup: spec => new _MenuItemGroup.MenuItemGroupBuilder(spec),
  menuItemsFromInitialValueTemplateItems: _InitialValueTemplateItem.menuItemsFromInitialValueTemplateItems,
  documentList: spec => new _DocumentList.DocumentListBuilder(spec),
  documentListItem: spec => new _DocumentListItem.DocumentListItemBuilder(spec),
  orderingMenuItem: ordering => (0, _MenuItem.getOrderingMenuItem)(ordering),
  orderingMenuItemsForType: type => (0, _MenuItem.getOrderingMenuItemsForSchemaType)(type),
  editor: spec => new _Editor.EditorBuilder(spec),
  editorWithInitialValueTemplate: _Editor.editorWithInitialValueTemplate,
  defaultInitialValueTemplateItems: _InitialValueTemplateItem.defaultInitialValueTemplateItems,
  initialValueTemplateItem: (templateId, parameters) => new _InitialValueTemplateItem.InitialValueTemplateItemBuilder({
    id: templateId,
    parameters,
    templateId
  }),
  component: spec => {
    return typeof spec === 'function' ? new _Component.ComponentBuilder().component(spec) : new _Component.ComponentBuilder(spec);
  },
  divider: () => ({
    id: (0, _uniqueId2.default)('__divider__'),
    type: 'divider'
  })
};
exports.StructureBuilder = StructureBuilder;

function hasIcon(schemaType) {
  if (!schemaType || typeof schemaType === 'string') {
    return false;
  }

  return Boolean(schemaType.icon);
}

function getDefaultStructure() {
  var items = (0, _documentTypeListItems.getDocumentTypeListItems)();
  return new _List.ListBuilder().id('__root__').title('Content').items(items).showIcons(items.some(item => hasIcon(item.getSchemaType())));
}
