// @sanity/components/lib/presence/PresenceList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _PresenceCircle = _interopRequireDefault(require("./PresenceCircle"));

var _PresenceList = _interopRequireDefault(require("./styles/PresenceList.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function filterMarkers(markers) {
  return markers.filter(marker => marker.type === 'presence');
}

class PresenceList extends _react.default.PureComponent {
  render() {
    var markers = filterMarkers(this.props.markers);
    return _react.default.createElement("ul", {
      className: _PresenceList.default.root
    }, markers.map(marker => {
      var user = marker.user;
      var imageUrl = user.imageUrl;
      var initials = user && user.displayName.match(/\b\w/g).join('') || '?';
      return _react.default.createElement("li", {
        key: marker.identity,
        className: _PresenceList.default.item
      }, _react.default.createElement("div", {
        className: _PresenceList.default.circle
      }, _react.default.createElement(_PresenceCircle.default, {
        imageUrl: imageUrl,
        color: marker.color,
        text: initials
      })), user.displayName);
    }));
  }

}

exports.default = PresenceList;

_defineProperty(PresenceList, "propTypes", {
  markers: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number, _propTypes.default.shape({
      _key: _propTypes.default.string
    })])),
    type: _propTypes.default.string,
    identity: _propTypes.default.string,
    session: _propTypes.default.string
  }))
});

_defineProperty(PresenceList, "defaultProps", {
  markers: []
});
