// @sanity/desk-tool/lib/pane/Editor/EditForm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _dateFns = require("date-fns");

var _formBuilder = _interopRequireDefault(require("part:@sanity/form-builder"));

var _EditorStatusBadge = _interopRequireDefault(require("../EditorStatusBadge"));

var _TimeAgo = _interopRequireDefault(require("../../components/TimeAgo"));

var _Editor = _interopRequireDefault(require("../styles/Editor.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var preventDefault = ev => ev.preventDefault();

class EditForm extends _react.default.PureComponent {
  render() {
    var _this$props = this.props,
        draft = _this$props.draft,
        published = _this$props.published,
        filterField = _this$props.filterField,
        focusPath = _this$props.focusPath,
        initialValue = _this$props.initialValue,
        isLiveEditEnabled = _this$props.isLiveEditEnabled,
        markers = _this$props.markers,
        onBlur = _this$props.onBlur,
        onChange = _this$props.onChange,
        onFocus = _this$props.onFocus,
        onShowHistory = _this$props.onShowHistory,
        patchChannel = _this$props.patchChannel,
        readOnly = _this$props.readOnly,
        schema = _this$props.schema,
        type = _this$props.type;
    var value = draft || published || initialValue;
    return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
      className: _Editor.default.top
    }, _react.default.createElement("span", {
      className: _Editor.default.statusBadges
    }, _react.default.createElement(_EditorStatusBadge.default, {
      liveEdit: isLiveEditEnabled,
      isDraft: !!draft,
      isPublished: !!published,
      title: published && "Published ".concat((0, _dateFns.distanceInWordsToNow)(published._updatedAt, {
        addSuffix: true
      }))
    })), value && value._updatedAt && _react.default.createElement("span", {
      className: _Editor.default.editedTimeClickable,
      onClick: onShowHistory
    }, 'Updated ', _react.default.createElement(_TimeAgo.default, {
      time: value._updatedAt
    }))), _react.default.createElement("form", {
      className: _Editor.default.editor,
      onSubmit: preventDefault,
      id: "Sanity_Default_DeskTool_Editor_ScrollContainer"
    }, _react.default.createElement(_formBuilder.default, {
      schema: schema,
      patchChannel: patchChannel,
      value: value || {
        _type: type
      },
      type: type,
      filterField: filterField,
      readOnly: readOnly,
      onBlur: onBlur,
      onFocus: onFocus,
      focusPath: focusPath,
      onChange: onChange,
      markers: markers
    })));
  }

}

exports.default = EditForm;

_defineProperty(EditForm, "propTypes", {
  draft: _propTypes.default.object,
  filterField: _propTypes.default.func.isRequired,
  focusPath: _propTypes.default.array.isRequired,
  isLiveEditEnabled: _propTypes.default.bool,
  markers: _propTypes.default.arrayOf(_propTypes.default.shape({
    path: _propTypes.default.array
  })).isRequired,
  onBlur: _propTypes.default.func.isRequired,
  onChange: _propTypes.default.func.isRequired,
  onFocus: _propTypes.default.func.isRequired,
  onShowHistory: _propTypes.default.func.isRequired,
  patchChannel: _propTypes.default.object.isRequired,
  published: _propTypes.default.object,
  initialValue: _propTypes.default.object,
  readOnly: _propTypes.default.bool.isRequired,
  schema: _propTypes.default.object.isRequired,
  type: _propTypes.default.object.isRequired
});
